hopsworks-api/src/main/java/io/hops/hopsworks/api/jwt/JWTResource.java

```diff
@@ -64,10 +77,55 @@ public Response createToken(JWTRequestDTO jWTRequestDTO) throws NoSuchAlgorithmE
   @Path("/renew")
   public Response renewToken(JsonWebTokenDTO jsonWebTokenDTO) throws SigningKeyNotFoundException, NotRenewableException,
       InvalidationException {
-    JWTResponseDTO jWTResponseDTO = jWTHelper.renewToken(jsonWebTokenDTO);
+    JWTResponseDTO jWTResponseDTO = jWTHelper.renewToken(jsonWebTokenDTO, true, new HashMap<>(3));
     return Response.ok().entity(jWTResponseDTO).build();
   }
 
+  @PUT
+  @Path("/renew/service")
+  public Response renewServiceToken(JsonWebTokenDTO jwt, @Context HttpServletRequest request)
+      throws HopsSecurityException {
+    // This token should be the one-time renewal token
+    String token = jWTHelper.getAuthToken(request);
+    Users user = jWTHelper.getUserPrincipal(request);
+    if (user == null) {
+      DecodedJWT decodedJWT = JWT.decode(token);
```

Comment:
Where do we check that the token has not been invalidated?

Reply:
In `JWTFilter`